### ISSUE TITLE
Use mps on kubernetes

### ISSUE THREAD

**Issue by somelaoda** [NVIDIA/k8s-device-plugin#467]
I'm trying to use mps service on kubernetes with nvidia-docker
```
Docker version 19.03.13,
nvidia-driver  495.44
cuda 11.5
image ：ngc tensorflow:21.11
```

Now I set nvidia-cuda-mps-control on the host machine，also hostIPC and hostPID  has been set when  nvidia-docker  startup.

Now the process in container can found nvidia-cuda-mps-control process, but the process memory limit is not in effect，No matter what I use 

$export CUDA_MPS_PINNED_DEVICE_MEM_LIMIT=”0=1G,1=512MB” 
or
set_default_device_pinned_mem_limit

how can I  make MPS work correctly across multiple containers?

**Comment by ettelr** [comment]
Any update on this thread? is anyone using MPS in kubernetes

**Comment by prattcmp** [comment]
Strange that this has been ignored for so long...

**Comment by elezar** [comment]
This is something that is under active development. We don't have a concrete release date yet, but are targetting the first quarter of 2024.

**Comment by prattcmp** [comment]
> This is something that is under active development. We don't have a concrete release date yet, but are targetting the first quarter of 2024.

2024Q1 would be great, even in a beta version.

**Comment by klueska** [comment]
We just released an RC for the next version of the k8s-device-plugin with support for MPS:
https://github.com/NVIDIA/k8s-device-plugin/tree/v0.15.0-rc.1?tab=readme-ov-file#with-cuda-mps

We would appreciate people to try this out and give any feedback you have before the final release in a few weeks.

**Comment by ettelr** [comment]
@klueska I see in the link you sent the following note:
Note: Sharing with MPS is currently not supported on devices with MIG enabled.
Is it planned to be  supported on GPUs that are not mig enabled (like L40, L40S) ? if yes would it be closely?

**Comment by hrbasic** [comment]
Hey, I'm trying to deploy `v0.15.0-rc.1` version, but I'm getting error: 
```
I0228 15:18:51.597975      31 main.go:279] Retrieving plugins.
I0228 15:18:51.598008      31 factory.go:104] Detected NVML platform: found NVML library
I0228 15:18:51.598047      31 factory.go:104] Detected non-Tegra platform: /sys/devices/soc0/family file not found
E0228 15:18:51.619657      31 main.go:301] Failed to start plugin: error waiting for MPS daemon: error checking MPS daemon health: failed to send command to MPS daemon: exit status 1
I0228 15:18:51.619670      31 main.go:208] Failed to start one or more plugins. Retrying in 30s...
```

Running with config:: 
```
{
  "version": "v1",
  "flags": {
    "migStrategy": "none",
    "failOnInitError": true,
    "mpsRoot": "/run/nvidia/mps",
    "nvidiaDriverRoot": "/",
    "gdsEnabled": false,
    "mofedEnabled": false,
    "plugin": {
      "passDeviceSpecs": false,
      "deviceListStrategy": [
        "envvar"
      ],
      "deviceIDStrategy": "uuid",
      "cdiAnnotationPrefix": "cdi.k8s.io/",
      "nvidiaCTKPath": "/usr/bin/nvidia-ctk",
      "containerDriverRoot": "/driver-root"
    }
  },
  "resources": {
    "gpus": [
      {
        "pattern": "*",
        "name": "nvidia.com/gpu"
      }
    ]
  },
  "sharing": {
    "timeSlicing": {},
    "mps": {
      "renameByDefault": true,
      "resources": [
        {
          "name": "nvidia.com/gpu",
          "rename": "nvidia.com/gpu.shared",
          "devices": "all",
          "replicas": 10
        }
      ]
    }
  }
}
```
All pods are up and running: 
```
gpu-feature-discovery-x59tw                                   2/2     Running     0          30m
gpu-operator-5bd8fb6df5-r2jrq                                 1/1     Running     0          103m
gpu-operator-node-feature-discovery-gc-78b479ccc6-kf8nk       1/1     Running     0          103m
gpu-operator-node-feature-discovery-master-569bfcd8bc-z6whl   1/1     Running     0          103m
gpu-operator-node-feature-discovery-worker-4bnwr              1/1     Running     0          103m
gpu-operator-node-feature-discovery-worker-d5cmh              1/1     Running     0          103m
gpu-operator-node-feature-discovery-worker-fc4vs              1/1     Running     0          103m
gpu-operator-node-feature-discovery-worker-ktsl9              1/1     Running     0          103m
gpu-operator-node-feature-discovery-worker-lm2gv              1/1     Running     0          103m
gpu-operator-node-feature-discovery-worker-mhmjv              1/1     Running     0          103m
gpu-operator-node-feature-discovery-worker-w4mgz              1/1     Running     0          103m
nvidia-container-toolkit-daemonset-qj5p6                      1/1     Running     0          100m
nvidia-cuda-validator-slg25                                   0/1     Completed   0          95m
nvidia-dcgm-exporter-kr84r                                    1/1     Running     0          100m
nvidia-device-plugin-2svlb                                    2/2     Running     0          30m
nvidia-driver-daemonset-sznx9                                 1/1     Running     0          103m
nvidia-mig-manager-78hrq                                      1/1     Running     0          100m
nvidia-operator-validator-nhz56                               1/1     Running     0          100m
```

I'm using `535.154.05` driver deployed with gpu-operator on Rocky 8.9. Any idea what could be the root cause?

**Comment by hrbasic** [comment]
Thanks for answer, I've already deployed this separately: 
**gpu-operator**
```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
metadata:
  name: gpu-operator

resources:
  - namespace.yaml

namespace: gpu-operator

helmCharts:
  - name: gpu-operator
    repo: https://nvidia.github.io/gpu-operator
    releaseName: gpu-operator
    namespace: gpu-operator
    valuesFile: values.yaml
    version: 23.9.1
```
values for operator:
```
driver:
  repository: my-repo.com/nvidia
  version: 535.154.05
  imagePullPolicy: Always
  imagePullSecrets: 
    - image-pull-secret
  useOpenKernelModules: true

gfd:
  enabled: false

mig:
  strategy: "none"
          
operator:
  imagePullSecrets: 
    - image-pull-secret

devicePlugin:
  enabled: false
```

**device-plugin**
```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
metadata:
  name: nvidia-device-plugin

namespace: gpu-operator

helmCharts:
  - name: nvidia-device-plugin
    repo: https://nvidia.github.io/k8s-device-plugin
    releaseName: nvidia-device-plugin
    namespace: gpu-operator
    valuesFile: values.yaml
    version: 0.15.0-rc.1
```

values for plugin: 
```
config:
  default: default
  map:
    default: |-
      version: v1
      flags:
        migStrategy: none
      sharing:
        mps:
          renameByDefault: true
          resources:
          - name: nvidia.com/gpu
            replicas: 10
```
But I'll try this once more on new node,  since it's deployed on node which already had "old" drivers installed by operator.

**Comment by klueska** [comment]
OK, that should work then (it's the same way I have tested things locally). Note that there may be a few transient failures in the plugin while it waits for the MPS daemonset to start up (because it won't come online until GFD has applied a label indicating that it should be there).

**Comment by hrbasic** [comment]
Just wanted to inform you that I successfully configured and deployed the device-plugin with MPS. I disabled `NFD` in the gpu-operator helm chart and enabled it in the device-plugin installation. Additionally, I had to restart\delete the `nvidia-device-plugin-gpu-feature-discovery` pod. I assume that restart was necessary since I've installed both helm charts simultaneously, or maybe it would eventually applied labels like you mentioned. Thanks for help, I'll keep you informed if any issues arise during the testing phase.

**Comment by igorgad** [comment]
Hello. I can confirm that installing the device-plugin version 0.15.0-rc.1 alongside gpu-operator works with the following procedure. 

1. Install gpu-operator with nvdp and nfd
2. upgrade to disable nvdp and nfd
3. Install nvdp with gfd enabled

One slight problem I'm facing is that it does segfault if a pod has `/dev/shm` mounted and try to allocate GPU memory, like in the following example. Is there a workaround to avoid using `/dev/shm` for the mps daemon communication?

```
apiVersion: v1
kind: Pod
metadata:
  name: testshm
spec:
  volumes:
    - emptyDir:
        medium: Memory
      name: shared-mem
  containers:
  - name: testshm
    image: nvidia/cuda:12.3.1-base-ubuntu20.04
    command: ["tail", "-f", "/dev/null"]
    volumeMounts:
      - mountPath: /dev/shm
        name: shared-mem
    resources:
      limits:
        nvidia.com/gpu: 1
```

Thanks!

**Comment by cdesiniotis** [comment]
@igorgad you do not need to manually mount `/dev/shm` in your pod spec. The device-plugin, as part of its AllocateResponse, will make sure all the entities required for MPS get included in the container.  Can you verify your example pod works when you remove the `shared-mem` volumeMount?

**Comment by elezar** [comment]
> @igorgad you do not need to manually mount `/dev/shm` in your pod spec. The device-plugin, as part of its AllocateResponse, will make sure all the entities required for MPS get included in the container. Can you verify your example pod works when you remove the `shared-mem` volumeMount?

To clarify: using MPS does require a `/dev/shm` to be set up and this is used by the MPS Control Daemon to allow for communication. The infrastructure added to the device plugin to support MPS automatically creates a tmpfs and mounts it at /dev/shm for any containers that require MPS. This means that the *additional* `/dev/shm` that you are requesting is overriding this `/dev/shm` that contians the information controlled by the MPS control daemon -- causing the segfaults.

**Comment by igorgad** [comment]
Hey @cdesiniotis and @elezar, thanks for clarifying it.

I can confirm that it works properly without the `shared-mem` volume mounted on the pod. However,  it's common to mount a memory-backed volume on `/dev/shm` to increase the amount of shared memory available to python multiprocessing and pytorch dataloaders. The tmpfs mounted to `/dev/shm` by the device plugin has 64MB, which is too small for many workloads.

**Comment by elezar** [comment]
We have an issue to track making the shm size configurable. Would this be able to address your use case? What are typical values for the shared memory size?

**Comment by igorgad** [comment]
Yep. Sounds good.
I generally set the shared memory size of the pod to the amount of memory requested by the pod. But I guess that's not feasible in the device plugin context. Therefore, I would say that 8GB should be enough for most workloads.

**Comment by klueska** [comment]
Future versions of MPS will not depend on /dev/shm at all, making the need to inject /dev/shm unnecessary. Until then (meaning on any existing driver) this issue will continue to persist.

**Comment by ettelr** [comment]
> Future versions of MPS will not depend on /dev/shm at all, making the need to inject /dev/shm unnecessary. Until then (meaning on any existing driver) this issue will continue to persist.

this means that currently torch workloads are not runnable an a system with nvidia mps deployed via device plugin
since almost every torch workload will need more than the amount that's injected currently by chart

**Comment by ZYWNB666** [comment]
If I want to use Time-Slicing in k8s, do I need to enable mps on node hosts?

**Comment by ettelr** [comment]
> If I want to use Time-Slicing in k8s, do I need to enable mps on node hosts?



> @ettelr we have an action item to allow the size of the `/dev/shm` that is created to be specified as part of the deployment. Would this work for your use cases?

Hi @elezar , @klueska 

is there any update on shm item?
is it already configurable from chart?  is it still required for mps daemon ?

**Comment by ettelr** [comment]
> @ettelr we have an action item to allow the size of the `/dev/shm` that is created to be specified as part of the deployment. Would this work for your use cases?

Hi any update on mps shm configurable ? we cannot use it like this each of workload is using different shm

**Comment by ettelr** [comment]
> > @ettelr we have an action item to allow the size of the `/dev/shm` that is created to be specified as part of the deployment. Would this work for your use cases?
> 
> Hi any update on mps shm configurable ? we cannot use it like this each of workload is using different shm

@klueska  @elezar

**Comment by zhiyxu** [comment]
> > @igorgad you do not need to manually mount `/dev/shm` in your pod spec. The device-plugin, as part of its AllocateResponse, will make sure all the entities required for MPS get included in the container. Can you verify your example pod works when you remove the `shared-mem` volumeMount?
> 
> To clarify: using MPS does require a `/dev/shm` to be set up and this is used by the MPS Control Daemon to allow for communication. The infrastructure added to the device plugin to support MPS automatically creates a tmpfs and mounts it at /dev/shm for any containers that require MPS. This means that the _additional_ `/dev/shm` that you are requesting is overriding this `/dev/shm` that contians the information controlled by the MPS control daemon -- causing the segfaults.

@elezar Hi, I found a file named cuda.shm.0.xx.1 in /dev/shm (where xx varies for each file), and this file is required when using MPS. The file is very small, only 4k in size. Detailed information is provided below. Do you know the specific content and purpose of this file? 

```
$ stat /dev/shm/cuda.shm.0.xx.1
  File: /dev/shm/cuda.shm.0.xx.1
  Size: 4096      	Blocks: 8          IO Block: 4096   regular file
Device: 15h/21d	Inode: 1008193229  Links: 1
Access: (0600/-rw-------)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2024-08-09 14:57:42.483693717 +0800
Modify: 2024-08-09 14:29:23.214025824 +0800
Change: 2024-08-09 14:29:23.214025824 +0800
 Birth: -
```

Additionally, this path seems to be hardcoded, as the file must be located in /dev/shm. Is there any way to configure the location of this file?

**Comment by ettelr** [comment]
@klueska  @elezar 

- Is there any plans for the configurable shm ? 
- Is there an option to enable to give a string value for renameByDefault so we can give our own custom name  (helpful in case of multiple configs for several nodes)
